keps/sig-scheduling/3521-pod-schedule-readiness/README.md

```diff
@@ -0,0 +1,1064 @@
+
+- **New field disabled in Alpha but not scheduler extension:** In Alpha, the new Pod field is disabled
+by default. However, the scheduler's extension point is activated no matter the feature gate is enabled
+or not. This enables scheduler plugin developers to tryout the feature even in Alpha, by crafting
+different enqueue plugins and wire with custom fields or conditions.
+
+- **New column in kubectl:** To provide better UX, we're going to add a column "scheduling paused"
+to the output of `kubectl get pod` to indicate whether it's scheduling-paused or not.
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+- The scheduler doesn't actively clear a Pod's `schedulingGates` field. This means if some controller
+is ill-implemented, some Pods may stay in Pending state incorrectly. If you noticed a Pod stays
+in Pending state for a long time and it carries non-nil `schedulingGates`, you may find out which
+component owns the gate(s) (via `.spec.managedFields`) and report the symptom to the component owner.
+
+- Faulty controllers may forget to remove the Pod's `schedulingGates`, and hence results in
+a large number of unschedulable Pods. In Alpha, we don't limit the number of unschedulable Pods caused
+by potential faulty controllers. We will evaluate necessary options in the future to mitigate
+potential abuse.
+
+- End-users may be confused by no scheduling events in the output of `kubectl describe pod xyz`. We
+will provide detailed documentation, along with metrics, tooling (kubectl) and(or) events.
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+### API
+
+A new API field `SchedulingGates` will be added to Pod's spec:
+
+```go
+type PodSpec struct {
+    // Each scheduling gate represents a particular scenario the scheduling is blocked upon.
+    // Scheduling is triggered only when SchedulingGates is empty.
+    // In the future, we may impose permission mechanics to restrict which controller can mutate
+    // which scheduling gate. It's dependent on a yet-to-be-implemented fined-grained
+    // permission (https://docs.google.com/document/d/11g9nnoRFcOoeNJDUGAWjlKthowEVM3YGrJA3gLzhpf4)
+    // and needs to be consistent with how finalizes/liens work today.
+    SchedulingGates []PodSchedulingGate
+}
+
+type PodSchedulingGate struct {
+    // Name of the scheduling gate.
+    Name string
```

Comment:
Clarify here that name must be unique among gates?  You say map earlier, but we don't clearly describe here that this must be unique among other gates.

Reply:
Updated.